easybuild/framework/easyconfig/easyconfig.py

```diff
@@ -134,6 +134,60 @@ def cache_aware_func(toolchain):
     return cache_aware_func
 
 
+def det_subtoolchain_version(current_tc, subtoolchain_name, optional_toolchains, cands):
+    """
+    Returns unique version for subtoolchain, in tc dict.
+    If there is no unique version:
+    * use '' for dummy, if dummy is not skipped.
+    * GCCcore can be downgraded to dummy.
+    * return None for skipped subtoolchains, that is,
+      optional toolchains or dummy without add_dummy_to_minimal_toolchains.
+    * in all other cases, raises an exception.
+    """
+    current_tc_name, current_tc_version = current_tc['name'], current_tc['version']
+
+    uniq_subtc_versions = set([subtc['version'] for subtc in cands if subtc['name'] == subtoolchain_name])
+
+    if len(uniq_subtc_versions) == 1:
+        subtoolchain_version = list(uniq_subtc_versions)[0]
+    elif len(uniq_subtc_versions) == 0:
+        # only retain GCCcore as subtoolchain if version was found
+        if subtoolchain_name == GCCcore.NAME:
+            _log.info("No version found for %s; assuming legacy toolchain and skipping it as subtoolchain.",
+                      subtoolchain_name)
+            # downgrade to dummy
+            subtoolchain_name = GCCcore.SUBTOOLCHAIN
+            subtoolchain_version = ''
+        # dummy toolchain: bottom of the hierarchy
+        elif subtoolchain_name == DUMMY_TOOLCHAIN_NAME:
+            subtoolchain_version = ''
+        else:
+            is_optional_tc = subtoolchain_name in optional_toolchains or current_tc_name in optional_toolchains
+            if is_optional_tc and robot_find_easyconfig(subtoolchain_name, current_tc_version):
+                # special case: optionally find e.g. golf/1.4.10 for goolf/1.4.10 even if it is not in
+                # the module dependencies. This is only allowed for and inside optional subtoolchains.
+                subtoolchain_version = current_tc_version
+            elif subtoolchain_name in optional_toolchains:
+                # skip if the subtoolchain considered now is optional
+                return None
+            else:
+                raise EasyBuildError("No version found for subtoolchain %s in dependencies of %s",
+                                     subtoolchain_name, current_tc_name)
+    elif subtoolchain_name == DUMMY_TOOLCHAIN_NAME:
+        # Don't care about multiple versions of dummy
+        _log.info("Ignoring multiple versions of %s in toolchain hierarchy", DUMMY_TOOLCHAIN_NAME)
+        subtoolchain_version = ''
+    else:
+        raise EasyBuildError("Multiple versions of %s found in dependencies of toolchain %s: %s",
+                             subtoolchain_name, current_tc_name, ', '.join(sorted(uniq_subtc_versions)))
+
+    if subtoolchain_name == DUMMY_TOOLCHAIN_NAME and not build_option('add_dummy_to_minimal_toolchains'):
+        # skip dummy if add_dummy_to_minimal_toolchains is not set
+        return None
+
+    return {'name': subtoolchain_name, 'version': subtoolchain_version}
```

Comment:
@bartoldeman I would just return `subtoolchain_version` here? I don't really see the point in returning a dict value here...

Reply:
I can't because of the change of name from "GCCcore" to "dummy" for legacy toolchains. I did not want to remove that code.

I suspect the idea is that older toolchains had GCC directly above dummy, but the gcc toolchain has GCCcore as the subtoolchain which does not exist in those situations.

Reply:
Ah, OK, I figured there must have been a good reason.

Then: i) mention that in a comment, explain why you need to return the subtoolchain name too, ii) just make this a tuple `(subtoolchain_name, subtoolchain_version)`.

Reply:
There are some ways around this of course, but e.g. moving the GCCcore logic back to the main function needs another check for dummy there too then, which is awkward...

Reply:
No, it's fine like this, just needs a comment since it doesn't make sense unless you take the whole function into account.

Reply:
In the end I found a more elegant way by declaring the gcccore toolchain as optional. It still needs some special logic in the main loop, but det_subtoolchain_version does exactly that now and nothing more.

Next will be tests for that but probably on Monday :)